File: RaftSwift/Sources/RaftCore/Node/RaftNodeApplication.swift
/// Protocol for a Raft node application
public protocol RaftNodeApplication: Sendable {
    /// The own peer
    var ownPeer: Peer { get }

    /// The list of peers
    var peers: [Peer] { get }

    /// The persistence layer
    var persistence: any RaftNodePersistence { get }

    /// Whether to collect metrics
    var collectMetrics: Bool { get }

    /// Whether to use a manual lock
    var useManualLock: Bool { get }

    /// Initializes the server
    /// - Parameters:
    ///   - ownPeer: The own peer
    ///   - peers: The list of peers
    ///   - persistence: The persistence layer
    ///   - collectMetrics: Whether to collect metrics
    ///   - useManualLock: Whether to use a manual lock
    init(
        ownPeer: Peer,
        peers: [Peer],
        persistence: any RaftNodePersistence,
        collectMetrics: Bool,
        useManualLock: Bool,
    )

    /// Starts the node
    func serve() async throws
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/core/node/RaftNodeApplication.kt
package raft.core.node

import raft.core.utils.types.Peer

/**
 * Abstract class for a Raft node application.
 * This class provides the basic functionality for a Raft node application.
 */
abstract class RaftNodeApplication {
    /**
     * The own peer.
     */
    val ownPeer: Peer
    /**
     * The list of peers.
     */
    val peers: MutableList<Peer>
    /**
     * The persistence layer.
     */
    val persistence: RaftNodePersistence
    /**
     * Whether to collect metrics.
     */
    val collectMetrics: Boolean

    /**
     * Constructor for a Raft node application.
     *
     * @param ownPeer The own peer.
     * @param peers The list of peers.
     * @param persistence The persistence layer.
     */
    constructor(ownPeer: Peer, peers: MutableList<Peer>, persistence: RaftNodePersistence, collectMetrics: Boolean) {
        this.ownPeer = ownPeer
        this.peers = peers
        this.persistence = persistence
        this.collectMetrics = collectMetrics
    }

    /**
     * Starts the Raft node application.
     */
    abstract fun serve()
}